### ISSUE TITLE
incompatible with python 3.10

### ISSUE THREAD

**Issue by 2bndy5** [mozilla/sphinx-js#186]
This is the third time I've run into this error using python 3.10.
```
Could not import extension sphinx_js (exception: cannot import name 'Mapping' from 'collections' (C:\Users\%user_name%\AppData\Local\Programs\Python\Python310\lib\collections\__init__.py))
```

Apparently, `Mapping` no longer exists (as an alias) in the `collections` module.

The backward compatible fix that I've been suggesting to other python devs hitting the same problem is
```py
try:
    from collections.abc import Mapping
except ImportError:
    # TODO: remove when dropping python 2.7
    from collections import Mapping
```
 However, I can't narrow down where exactly this would be applied. When doing a search for the keyword collections, I only come up with 

https://github.com/mozilla/sphinx-js/blob/a1ee45651f14bba63a4e2e370e9acadc847c017b/sphinx_js/jsdoc.py#L8

and also the same line in a file called doclets.py (I can't find that file here on github though).

I think `defaultdict` implicitly looks for Mapping from collections, so I have no idea about how to solve this without taking this upstream (trying to figure out [python's bug tracker](https://bugs.python.org/) seems like too much hassle).

**Comment by 2bndy5** [comment]
Found this comment in the python issue tracker:
> According to https://docs.python.org/3.9/library/collections.html, aliases like collections.Mapping were removed from the collections namespace in 3.10.

**Comment by 2bndy5** [comment]
I think this is a [sphinx-js dependency](https://github.com/mozilla/sphinx-js/blob/a1ee45651f14bba63a4e2e370e9acadc847c017b/setup.py#L17) problem (concerning parsimonious pkg). 

<details><summary>Full trace back from <code>sphinx-build . _build</code></summary>

```
Original exception:
Traceback (most recent call last):
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx\registry.py", line 429, in load_extension
    mod = import_module(extname)
  File "C:\Users\%user_name%\AppData\Local\Programs\Python\Python310\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx_js\__init__.py", line 5, in <module>
    from .directives import (auto_class_directive_bound_to_app,
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx_js\directives.py", line 13, in <module>
    from .renderers import AutoFunctionRenderer, AutoClassRenderer, AutoAttributeRenderer
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx_js\renderers.py", line 12, in <module>
    from .parsers import PathVisitor
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx_js\parsers.py", line 3, in <module> 
    from parsimonious import Grammar, NodeVisitor
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\parsimonious\__init__.py", line 8, in <module>
    from parsimonious.grammar import Grammar, TokenGrammar
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\parsimonious\grammar.py", line 8, in <module>
    from collections import Mapping
ImportError: cannot import name 'Mapping' from 'collections' (C:\Users\%user_name%\AppData\Local\Programs\Python\Python310\lib\collections\__init__.py)


Traceback (most recent call last):
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx\registry.py", line 429, in load_extension
    mod = import_module(extname)
  File "C:\Users\%user_name%\AppData\Local\Programs\Python\Python310\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx_js\__init__.py", line 5, in <module>
    from .directives import (auto_class_directive_bound_to_app,
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx_js\directives.py", line 13, in <module>
    from .renderers import AutoFunctionRenderer, AutoClassRenderer, AutoAttributeRenderer
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx_js\renderers.py", line 12, in <module>
    from .parsers import PathVisitor
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx_js\parsers.py", line 3, in <module> 
    from parsimonious import Grammar, NodeVisitor
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\parsimonious\__init__.py", line 8, in <module>
    from parsimonious.grammar import Grammar, TokenGrammar
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\parsimonious\grammar.py", line 8, in <module>
    from collections import Mapping
ImportError: cannot import name 'Mapping' from 'collections' (C:\Users\%user_name%\AppData\Local\Programs\Python\Python310\lib\collections\__init__.py)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx\cmd\build.py", line 276, in build_main
    app = Sphinx(args.sourcedir, args.confdir, args.outputdir,
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx\application.py", line 237, in __init__
    self.setup_extension(extension)
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx\application.py", line 394, in setup_extension
    self.registry.load_extension(self, extname)
  File "C:\Users\%user_name%\Documents\GitHub\venv\lib\site-packages\sphinx\registry.py", line 432, in load_extension
    raise ExtensionError(__('Could not import extension %s') % extname,
sphinx.errors.ExtensionError: Could not import extension sphinx_js (exception: cannot import name 'Mapping' from 'collections' (C:\Users\%user_name%\AppData\Local\Programs\Python\Python310\lib\collections\__init__.py))

Extension error:
Could not import extension sphinx_js (exception: cannot import name 'Mapping' from 'collections' (C:\Users\%user_name%\AppData\Local\Programs\Python\Python310\lib\collections\__init__.py))
```
</details>

I was able to fix this by changing
```py
from collections import Mapping
```
to
```py
from collections.abc import Mapping
```
in the _site-packages\parsimonious\grammar.py_

**Comment by dhdaines** [comment]
The dependency on parsimonious, which is fixed at 0.7, needs to be updated ... newer versions seem to work just fine for me, but perhaps that should be validated.

**Comment by 2bndy5** [comment]
I no longer have this installed as I was doing a trial run. Glad to close this out if you can verify with a fresh venv. Looks like the dependencies don't need to be updated here:
https://github.com/mozilla/sphinx-js/blob/1dd066bedd0b30d11697b64e395b6e6eb87adc39/setup.py#L17

Unless I read that comment wrong.

**Comment by dhdaines** [comment]
That's actually exactly where they need to be updated!  Just changing it to `parsimonious~=0.8.0` works fine for me... I'll submit a PR (with another small fix)

**Comment by 2bndy5** [comment]
Be sure to use [github's closing words](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) in the PR title or description, so this gets auto-closed when merged.

**Comment by 2bndy5** [comment]
should also be addressed with #184

**Comment by willkg** [comment]
This was fixed in 84ee24d2.

**Comment by sylvestre** [comment]
@willkg would it be possible to publish a new version on pip ? :)
thanks
